src/providers/MeetingProvider/types.ts

```diff
@@ -13,6 +14,7 @@ export enum DevicePermissionStatus {
 export interface MeetingJoinData {
   meetingInfo: any;
   attendeeInfo: any;
+  deviceLabels?: DeviceLabels | DeviceLabelTrigger;
```

Comment:
Could we not add here? deveiceLabels seems not related to meeting join data at all? You should put it as another optional parameter.
Also should we pass it through join? I think a better place would be meetingConfig when you first create meetingManager.

Reply:
For the first question, I agree with you. I will move the `deviceLabels` out of the `MeetingJoinData` and just make it as an optional parameter.

For the second question, I think the key point here is the timing of setting device permission request.

We could pass the `DeviceLabel` via `meetingCongfig` at the very beginning. But the `DeviceLabel` could be changed , right?
 
Let's say, we set the `AudioAndVideo` as default value in `meetingConfig`. So the default setting is for the **Attendee Mode**. When user wants to switch to the **Spectator mode**, they need do something to change the device permission setting before joining the meeting. So here is the timing. It's best to setup the device permission right before they joining the meeting. That's why I think it's a good idea to pass the parameter in the `meetingManager.join()`.

How do you think?

Reply:
If you agree that the timing before joining is a key point, and we must do something at this moment. Then the pass the `DeviceLabel` via `meetingConfig` seems redundant. Cause we need to setup the device permission before joining the meeting anyway.

Reply:
Sure we can just have one function to setDeviceLabelMode or something.

Reply:
I want to reiterate: the _browser_ prompts for permissions, and not all applications (a) need to enumerate devices, (b) need device labels, or (c) need permissions at all (e.g., Electron apps).

Nor do all applications have a linear procession through creating a meeting manager, joining a meeting, prompting for labels, etc. — the same window might join multiple calls, and permissions or labels are already a solved problem for subsequent sessions.

Reply:
@richnew10 Yes, so I think it's good to have the optional parameter in `meetingManager.join()` function to let use choose the preset settings for `None`, `Audio`, `Video` and `AudioAndVideo`, and expose the functions like below to control specific type of device, how do you think about this?
```ts
  skipDeviceProviderInit(deviceKind: DeviceKind): void {
    switch (deviceKind) {
      case DeviceKind.AudioInput:
        if (this.skipAudioInputProviderInit) {
          this.skipAudioInputProviderInit();
        }
        break;
      case DeviceKind.AudioOutput:
        if (this.skipAudioOutputProviderInit) {
          this.skipAudioOutputProviderInit();
        }
        break;
      case DeviceKind.VideoInput:
        if (this.skipVideoInputProviderInit) {
          this.skipVideoInputProviderInit();
        }
        break;
      default:
        break;
    }
  };

  invokeDeviceProviderInit(deviceKind: DeviceKind): void {
    switch (deviceKind) {
      case DeviceKind.AudioInput:
        if (this.invokeAudioInputProviderInit) {
          this.invokeAudioInputProviderInit();
        }
        break;
      case DeviceKind.AudioOutput:
        if (this.invokeAudioOutputProviderInit) {
          this.invokeAudioOutputProviderInit();
        }
        break;
      case DeviceKind.VideoInput:
        if (this.invokeVideoInputProviderInit) {
          this.invokeVideoInputProviderInit();
        }
        break;
      default:
        break;
    }
  };

  invokeDeviceLabelTrigger(deviceLabels: DeviceLabels): void {
    this.setupDeviceLabelTrigger(deviceLabels);

    switch (deviceLabels) {
      case DeviceLabels.AudioAndVideo:
        this.invokeDeviceProviderInit(DeviceKind.AudioInput);
        this.invokeDeviceProviderInit(DeviceKind.AudioOutput);
        this.invokeDeviceProviderInit(DeviceKind.VideoInput);
        break;
      case DeviceLabels.Audio:
        this.invokeDeviceProviderInit(DeviceKind.AudioInput);
        this.invokeDeviceProviderInit(DeviceKind.AudioOutput);
        break;
      case DeviceLabels.Video:
        this.invokeDeviceProviderInit(DeviceKind.VideoInput);
        break;
      default:
        break;
    }
  };
```

Reply:
I don't really understand the `skip*ProviderInit` functions, or why the name `skip` should be used. All you're really trying to do, right, is to control a particular hook that's run during initialization of this provider. It seems like the way to do that is a constructor or init argument. What am I missing?